File: NURBS_viewer/autogl/core/autogl_mp/sample_c/selection2d.c
/*  AutoGL $B!c$*<j7Z!d(B $B$N%5%s%W%k%W%m%0%i%`Nc(B */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <assert.h>

/* AutoGL $B!I$*$F$,$k!I$r;H$&$K$O(B,$B$3$N%X%C%@%U%!%$%k$r%$%s%/%k!<%I$9$k!#(B*/
#include <autogl.h>

/* AutoNOSim $B!I$*$?$N$7$_!I$r;H$&$K$O(B,$B$3$N%X%C%@%U%!%$%k$r%$%s%/%k!<%I$9$k!#(B*/
#include <autonosim.h>

/* $B%f!<%F%#%j%F%#$NFs<!85%$%Y%s%H=hM}$NMxMQ(B */
/* $B0z?t$J$I>\$7$$>pJs$O(B, autogl_utility2d.h$B$r;2>H!#(B*/
/* $B$J$*!"$3$l$O(Bautogl.h$B$+$i(Binclude$B$5$l$F$$$k!#(B*/



/* $B%S%e!<>e$KI=<($5$l$k%b%G%k$rI=8=$9$k$?$a$NJQ?t72(B */

/* $B8=:_A*Br$5$l$F$$$kD:E@$NHV9f(B */
static int SelectedEdgeId = -1;

/* $BD:E@$N:BI8(B */
static double Vertexs[6][2] = {
  { 10, 0 },
  { 10, 20 },
  { 0, 20 },
  { -10, 0 },
  { 0, -20 },
  { 30, -20 }
};



/* $B%S%e!<$N:FIA2h$N$?$a$N%3!<%k%P%C%/4X?t(B */
/* $B%S%e!<$,:FI=<($5$l$k$4$H$K8F$P$l$k!#(B */
static void RedrawView (void) 
{
  int i;
  int nTasks = AutoNOSim_NTasks ();
  int taskId = AutoNOSim_MyTaskId ();

  AutoGL_SetColor (0, 1, 0);  

  AutoGL_DrawLine (-50, 0, 0,
		   50, 0, 0); 
  AutoGL_DrawString (50, 0, 0, "X = 5O");  
  AutoGL_DrawString (-50, 0, 0, "X = -5O");  
  
  AutoGL_DrawLine (0, -50, 0,
		   0, 50, 0); 
  AutoGL_DrawString (0, 50, 0, "Y = 5O");  
  AutoGL_DrawString (0, -50, 0, "Y = -5O");  
    
  AutoGL_SetColor (1, 1, 1);
  AutoGL_DrawString (0, 0, 0, "O"); 

  for (i = 0; i < 6 - 1; i++) {
    if (i % nTasks != taskId) continue;

    if (i == SelectedEdgeId) {
      AutoGL_SetColor (1, 0, 0);
    } else {
      AutoGL_SetColor (1, 1, 1);
    }
    AutoGL_DrawLine (Vertexs[i][0], Vertexs[i][1], 0,
		     Vertexs[i + 1][0], Vertexs[i + 1][1], 0);
  }
}

/* $B%S%e!<$N%$%Y%s%H=hM}$?$a$N%3!<%k%P%C%/4X?t(B */
/* $B%S%e!<$K%$%Y%s%H$,H/@8$9$k$4$H$K8F$P$l$k!#(B */
static void HandleEvent (void)
{
  int event = AutoGL_GetViewEvent ();
  int nTasks = AutoNOSim_NTasks ();
  int taskId = AutoNOSim_MyTaskId ();

  /* $B$b$7!"%^%&%9%/%j%C%/$,$"$l$P(B */
  if (event == AUTOGL_EVENT_POINTING_DEVICE_RELEASE) {
    int dcX, dcY, dcZ;
    int toleranceDc = 10;
    int i;

    AutoGL_GetPointingDevicePositionDc 
      (&dcX, &dcY, &dcZ);

    AutoGL_StartSelection (dcX, dcY, toleranceDc);
    
    for (i = 0; i < 6 - 1; i++) {
      if (i % nTasks != taskId) continue;

      AutoGL_SetSelectionId (i);
      AutoGL_DrawLine (Vertexs[i][0], Vertexs[i][1], 0,
		       Vertexs[i + 1][0], Vertexs[i + 1][1], 0);
    }

    AutoGL_EndSelection ();
    SelectedEdgeId = AutoGL_GetSelectedId ();

    fprintf (stderr, " selected edge ID is %d \n", SelectedEdgeId);

    /* $B%S%e!<$r:FIA2h$9$k!#(B */
    AutoGL_DrawView ();
    /* $BEPO?$5$l$?%S%e!<$N:FIA2h4X?t!J$3$N>l9g$O(BRedrawView ()$B!K$,8F$P$l$k!#(B */

  } else if (event == AUTOGL_EVENT_KEY) {
    int keyChar = AutoGL_GetKeyChar ();

    if (keyChar == ' ') {
      fprintf (stderr, " hit space \n");
    } else {
      AutoGL_HandleDefaultKeyEventInMode2D ();
    }
  }
}



/* $B4X?t(BAutoGL_SetUp$B$O%f!<%6!<B&%W%m%0%i%`$4$H$KI,$:0l$DMQ0U$9$k$3$H!#(B*/
/* $B$3$3$G!"%3!<%k%P%C%/4X?t$d@)8fJQ?t$J$I$rEPO?$9$k!#(B*/
void AutoGL_SetUp (int argc, char *argv[]) 
{
  AutoGL_SetViewSize (70);   

  /* $B%S%e!<$N:FIA2h%3!<%k%P%C%/$NEPO?(B */
  AutoGL_SetViewRedrawCallback (RedrawView);   

  /* $B%S%e!<$N%$%Y%s%H=hM}%3!<%k%P%C%/$NEPO?(B */
  AutoGL_SetDefaultCallbackInMode2D (HandleEvent);
  /* $B%$%Y%s%H=hM}%3!<%k%P%C%/4X?t$H$7$F>e$K$"$k(BHandleEvent ()$B$r;XDj!#(B */
  /* $BFCJL$J%$%Y%s%H=hM}$,MW$i$J$1$l$P!"(BNULL$B$G$h$$!#(B*/

  /* $B=i4|>uBV$G$O!"%f!<%6!<Dj5A%$%Y%s%H=hM}$r9T$&!#(B */
  AutoGL_SetMode2D (AUTOGL_MODE_2D_OTHER); 

#if 0
  /* $B%$%Y%s%H=hM}Cf(B, $B%I%i%C%0$rM-8z$K$9$k!#(B*/
  AutoGL_EnableDragInMode2D ();
#endif

  /* $B%Q%M%k>e$KFs<!85%"%W%jMQ%$%Y%s%H=hM}5!G=$r=`Hw$9$k!#(B */
  AutoGL_SetPanelInMode2D ();
}

File: NURBS_viewer/autogl/sample_f/selection2d.f
c     
c     メインルーチン
c     
      implicit none

      real*8 vertexs(6, 2)
      integer selected_edge_id
      common vertexs, selected_edge_id

c     頂点の座標 
      vertexs(1, 1) = 10.0d0
      vertexs(1, 2) = 0.0d0

      vertexs(2, 1) = 10.0d0
      vertexs(2, 2) = 20.0d0

      vertexs(3, 1) = 0.0d0
      vertexs(3, 2) = 20.0d0

      vertexs(4, 1) = -10.0d0
      vertexs(4, 2) = 0.0d0

      vertexs(5, 1) = 0.0d0
      vertexs(5, 2) = -20.0d0

      vertexs(6, 1) = 30.0d0
      vertexs(6, 2) = -20.0d0

c     現在選択されている稜線の番号 
      selected_edge_id = -1



      call autogl_set_view_size(70.0d0)

c     イベント処理コールバック関数としてautogl_handle_eventを指定。 
      call autogl_set_default_callback_in_mode2d()

c     初期状態では、ユーザー定義イベント処理を行う。 
      call autogl_set_mode2d(0)

c     イベント処理中, ドラッグを有効にする。
      call autogl_enable_drag_in_mode2d()

c     パネル上に二次元アプリ用イベント処理機能を準備する。 
      call autogl_set_panel_in_mode2d()

c     AutoGLのメインループ
c     二度と戻らない
      call autogl_main()

      end



c     サブルーチンautoglredraw
c     ビューの再描画のためのコールバックサブルーチン 
c     ビューが再表示されるごとに呼ばれる。 
      subroutine autogl_redraw
      implicit none

      real*8 vertexs(6, 2)
      integer selected_edge_id
      common vertexs, selected_edge_id

      integer i

      call autogl_set_color(0.0d0, 1.0d0, 0.0d0) 

      call autogl_draw_line(
     c     -50.0d0, 0.0d0, 0.0d0,
     c     50.0d0, 0.0d0, 0.0d0)
      call autogl_draw_string(50.0d0, 0.0d0, 0.0d0, "X = 50") 
      call autogl_draw_string(-50.0d0, 0.0d0, 0.0d0, "X = -50") 
         
      call autogl_draw_line(
     c     0.0d0, -50.0d0, 0.0d0,
     c     0.0d0, 50.0d0, 0.0d0)
      call autogl_draw_string(0.0d0, 50.0d0, 0.0d0, "Y = 50") 
      call autogl_draw_string(0.0d0, -50.0d0, 0.0d0, "Y = -50") 

      call autogl_set_color(1.0d0, 1.0d0, 1.0d0) 
      call autogl_draw_string(0.0d0, 0.0d0, 0.0d0, "O")

      do i = 1, 6 - 1
         if (i .eq. selected_edge_id) then
            call autogl_set_color(1.0d0, 0.0d0, 0.0d0) 
         else
            call autogl_set_color(1.0d0, 1.0d0, 1.0d0) 
         endif
         call autogl_draw_line(
     c        vertexs(i, 1), vertexs(i, 2), 0.0d0,
     c        vertexs(i + 1, 1), vertexs(i + 1, 2), 0.0d0)
      enddo

      end



      subroutine autogl_handle_event
      implicit none

      real*8 vertexs(6, 2)
      integer selected_edge_id
      common vertexs, selected_edge_id

      integer event
      integer dc_x, dc_y, dc_z
      integer tolerance_dc
      integer i
      integer key_char
      
      call autogl_get_view_event(event)

c     もし、マウスクリックがあれば 
      if (event .eq. 2) then 
         
         tolerance_dc = 10

         call autogl_get_pointing_device_position_dc(
     c        dc_x, dc_y, dc_z)
         
         call autogl_start_selection(dc_x, dc_y, tolerance_dc)

         do i = 1, 6 - 1
            call autogl_set_selection_id(i)
            call autogl_draw_line(
     c           vertexs(i, 1), vertexs(i, 2), 0.0d0,
     c           vertexs(i + 1, 1), vertexs(i + 1, 2), 0.0d0)
         enddo
         
         call autogl_end_selection()
         call autogl_get_selected_id(
     c        selected_edge_id)
         
         write(*,*) ' selected edge ID is ', selected_edge_id

c     ビューを再描画する。 
         call autogl_draw_view()

c     もし、キーボード入力があれば 
      elseif (event .eq. 3) then
         call autogl_get_key_char(key_char)
         if (key_char .eq. ichar(' ')) then
            write(*,*) ' hit space '
         else
            call autogl_handle_default_key_event_in_mode2d()
         endif
      endif
      end



      subroutine autogl_idle
      end

      subroutine autogl_callback(id)
      implicit none
      integer id
      end
